File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/android/keymaster_ext.py
"""
Android Key Attestation extension parsing (OID 1.3.6.1.4.1.11129.2.1.17).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from asn1crypto import core

from ..cert_utils import get_extension_value

KEYMASTER_EXT_OID_DOTTED = "1.3.6.1.4.1.11129.2.1.17"

ANDROID_SECURITY_LEVEL_LABELS = {0: "Software", 1: "TrustedEnvironment", 2: "StrongBox"}


@dataclass
class KeyDescription:
    attestationVersion: int
    attestationSecurityLevel: int
    keyMintVersion: int
    keyMintSecurityLevel: int
    attestationChallenge: bytes


class _SecurityLevel(core.Integer):
    tag = 10


class _KeyDescriptionSequence(core.Sequence):
    _fields = [
        ("attestationVersion", core.Integer),
        ("attestationSecurityLevel", _SecurityLevel),
        ("keyMintVersion", core.Integer),
        ("keyMintSecurityLevel", _SecurityLevel),
        ("attestationChallenge", core.OctetString),
    ]


def parse_key_description(leaf_der: bytes) -> Optional[KeyDescription]:
    container = get_extension_value(leaf_der, KEYMASTER_EXT_OID_DOTTED)
    if container is None:
        return None
    try:
        fields = _KeyDescriptionSequence.load(container, strict=True)
        values = [fields[index].native for index in range(4)]
        if any(value < 0 for value in values):
            return None
        return KeyDescription(*values, attestationChallenge=fields["attestationChallenge"].native)
    except (ValueError, TypeError):
        return None


def extract_attestation_challenge_from_cert(leaf_der: bytes) -> Optional[bytes]:
    kd = parse_key_description(leaf_der)
    return kd.attestationChallenge if kd else None


def is_hardware_attestation_security_level(description: Optional[KeyDescription]) -> bool:
    return description is not None and description.attestationSecurityLevel in (1, 2)

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/android/KeymasterExt.java
package com.azure.ai.vision.face.deviceattestation.android;

import com.azure.ai.vision.face.deviceattestation.crypto.CertUtils;
import org.bouncycastle.asn1.ASN1Enumerated;
import org.bouncycastle.asn1.ASN1Integer;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Sequence;

/**
 * Android Key Attestation extension parsing (OID 1.3.6.1.4.1.11129.2.1.17).
 * Parses the five leading fields of the KeyDescription SEQUENCE; the
 * attestationChallenge must equal the session challengeHash.
 */
public final class KeymasterExt {
    public static final String KEYMASTER_EXT_OID = "1.3.6.1.4.1.11129.2.1.17";

    private static final int ATTESTATION_VERSION_INDEX = 0;
    private static final int ATTESTATION_SECURITY_LEVEL_INDEX = 1;
    private static final int KEY_MINT_VERSION_INDEX = 2;
    private static final int KEY_MINT_SECURITY_LEVEL_INDEX = 3;
    private static final int ATTESTATION_CHALLENGE_INDEX = 4;

    /** Decoded leading fields of the KeyDescription SEQUENCE. */
    public record KeyDescription(
            int attestationVersion,
            int attestationSecurityLevel,
            int keyMintVersion,
            int keyMintSecurityLevel,
            byte[] attestationChallenge) {
    }

    public static KeyDescription parseKeyDescription(byte[] leafDer) {
        byte[] container = CertUtils.getExtensionValue(leafDer, KEYMASTER_EXT_OID);
        if (container == null) {
            return null;
        }
        try {
            ASN1Sequence sequence = ASN1Sequence.getInstance(container);
            KeyDescription description = new KeyDescription(
                    ASN1Integer.getInstance(sequence.getObjectAt(ATTESTATION_VERSION_INDEX)).intValueExact(),
                    ASN1Enumerated.getInstance(sequence.getObjectAt(ATTESTATION_SECURITY_LEVEL_INDEX)).intValueExact(),
                    ASN1Integer.getInstance(sequence.getObjectAt(KEY_MINT_VERSION_INDEX)).intValueExact(),
                    ASN1Enumerated.getInstance(sequence.getObjectAt(KEY_MINT_SECURITY_LEVEL_INDEX)).intValueExact(),
                    ASN1OctetString.getInstance(sequence.getObjectAt(ATTESTATION_CHALLENGE_INDEX)).getOctets());
            return description.attestationVersion() < 0 || description.attestationSecurityLevel() < 0
                    || description.keyMintVersion() < 0 || description.keyMintSecurityLevel() < 0 ? null : description;
        } catch (RuntimeException e) {
            return null;
        }
    }

    public static byte[] extractAttestationChallengeFromCert(byte[] leafDer) {
        KeyDescription kd = parseKeyDescription(leafDer);
        return kd == null ? null : kd.attestationChallenge();
    }

    public static boolean isHardwareAttestationSecurityLevel(KeyDescription description) {
        return description != null
                && (description.attestationSecurityLevel() == 1 || description.attestationSecurityLevel() == 2);
    }

    private KeymasterExt() {
    }
}
